File: sdk/python/mashgate/resources/compliance.py
"""Tenant-scoped compliance alert lifecycle."""

from __future__ import annotations

from enum import Enum
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from mashgate.client import MashgateClient


class AlertStatus(str, Enum):
    OPEN = "ALERT_STATUS_OPEN"
    UNDER_REVIEW = "ALERT_STATUS_UNDER_REVIEW"
    RESOLVED = "ALERT_STATUS_RESOLVED"
    ESCALATED = "ALERT_STATUS_ESCALATED"
    SAR_FILED = "ALERT_STATUS_SAR_FILED"
    CLOSED = "ALERT_STATUS_CLOSED"


class AlertSeverity(str, Enum):
    LOW = "ALERT_SEVERITY_LOW"
    MEDIUM = "ALERT_SEVERITY_MEDIUM"
    HIGH = "ALERT_SEVERITY_HIGH"
    CRITICAL = "ALERT_SEVERITY_CRITICAL"


class AlertCategory(str, Enum):
    AML = "ALERT_CATEGORY_AML"
    SANCTIONS = "ALERT_CATEGORY_SANCTIONS"
    PEP = "ALERT_CATEGORY_PEP"
    FRAUD = "ALERT_CATEGORY_FRAUD"
    TRANSACTION = "ALERT_CATEGORY_TRANSACTION"
    KYC = "ALERT_CATEGORY_KYC"
    WATCHLIST = "ALERT_CATEGORY_WATCHLIST"


class ComplianceResource:
    def __init__(self, client: MashgateClient) -> None:
        self._c = client

    def raise_alert(
        self,
        *,
        subject_id: str,
        subject_type: str,
        category: AlertCategory | str,
        severity: AlertSeverity | str,
        source: str,
        source_ref: str,
        description: str,
        idempotency_key: str,
        evidence: list[dict[str, Any]] | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "tenantId": self._c.require_tenant_id(),
            "subjectId": subject_id,
            "subjectType": subject_type,
            "category": category,
            "severity": severity,
            "source": source,
            "sourceRef": source_ref,
            "description": description,
        }
        if evidence is not None:
            body["evidence"] = evidence
        return self._c.request(
            "POST",
            "/v1/compliance/alerts",
            body=body,
            extra_headers={"Idempotency-Key": idempotency_key},
        )

    def get(self, alert_id: str) -> dict[str, Any]:
        return self._c.request(
            "GET",
            f"/v1/compliance/alerts/{alert_id}",
            query={"tenant_id": self._c.require_tenant_id()},
        )

    def list(
        self,
        *,
        subject_id: str | None = None,
        status: AlertStatus | str | None = None,
        severity: AlertSeverity | str | None = None,
        limit: int | None = None,
        cursor: str | None = None,
    ) -> dict[str, Any]:
        return self._c.request(
            "GET",
            "/v1/compliance/alerts",
            query={
                "tenant_id": self._c.require_tenant_id(),
                "subject_id": subject_id,
                "status": status,
                "severity": severity,
                "limit": limit,
                "cursor": cursor,
            },
        )

    def resolve(self, alert_id: str, resolve_note: str) -> dict[str, Any]:
        return self._c.request(
            "POST",
            f"/v1/compliance/alerts/{alert_id}/resolve",
            body={
                "tenant_id": self._c.require_tenant_id(),
                "alert_id": alert_id,
                "resolve_note": resolve_note,
            },
        )

    def escalate(
        self, alert_id: str, *, escalated_to: str, escalate_note: str
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            f"/v1/compliance/alerts/{alert_id}/escalate",
            body={
                "tenant_id": self._c.require_tenant_id(),
                "alert_id": alert_id,
                "escalated_to": escalated_to,
                "escalate_note": escalate_note,
            },
        )

    def has_open_alerts(self, subject_id: str) -> bool:
        response = self.list(subject_id=subject_id, status=AlertStatus.OPEN, limit=1)
        return bool(response.get("alerts"))

File: sdk/typescript/src/resources/compliance.ts
import type { MashgateClient } from "../client.js";

export const AlertStatus = {
  Open: "ALERT_STATUS_OPEN",
  UnderReview: "ALERT_STATUS_UNDER_REVIEW",
  Resolved: "ALERT_STATUS_RESOLVED",
  Escalated: "ALERT_STATUS_ESCALATED",
  SARFiled: "ALERT_STATUS_SAR_FILED",
  Closed: "ALERT_STATUS_CLOSED",
} as const;
export type AlertStatus = (typeof AlertStatus)[keyof typeof AlertStatus];

export const AlertSeverity = {
  Low: "ALERT_SEVERITY_LOW",
  Medium: "ALERT_SEVERITY_MEDIUM",
  High: "ALERT_SEVERITY_HIGH",
  Critical: "ALERT_SEVERITY_CRITICAL",
} as const;
export type AlertSeverity = (typeof AlertSeverity)[keyof typeof AlertSeverity];

export const AlertCategory = {
  AML: "ALERT_CATEGORY_AML",
  Sanctions: "ALERT_CATEGORY_SANCTIONS",
  PEP: "ALERT_CATEGORY_PEP",
  Fraud: "ALERT_CATEGORY_FRAUD",
  Transaction: "ALERT_CATEGORY_TRANSACTION",
  KYC: "ALERT_CATEGORY_KYC",
  Watchlist: "ALERT_CATEGORY_WATCHLIST",
} as const;
export type AlertCategory = (typeof AlertCategory)[keyof typeof AlertCategory];

export interface AlertEvidence {
  evidenceType: string;
  referenceId: string;
  description: string;
  url: string;
}

export interface ComplianceAlert {
  alertId: string;
  tenantId: string;
  subjectId: string;
  subjectType: string;
  category: AlertCategory;
  severity: AlertSeverity;
  status: AlertStatus;
  source: string;
  sourceRef: string;
  description: string;
  evidence: AlertEvidence[];
  assignedTo: string;
  resolvedBy: string;
  resolveNote: string;
  escalatedTo: string;
  escalateNote: string;
  sarId: string;
  createdAt: string;
  updatedAt: string;
  dueAt?: string;
}

export interface RaiseComplianceAlert {
  subjectId: string;
  subjectType: string;
  category: AlertCategory;
  severity: AlertSeverity;
  source: string;
  sourceRef: string;
  description: string;
  evidence?: AlertEvidence[];
}

export interface ListComplianceAlertsQuery {
  subjectId?: string;
  status?: AlertStatus;
  severity?: AlertSeverity;
  limit?: number;
  cursor?: string;
}

export interface ListComplianceAlertsResponse {
  alerts: ComplianceAlert[];
  nextCursor?: string;
}

export class ComplianceResource {
  constructor(private readonly client: MashgateClient) {}

  raise(data: RaiseComplianceAlert, idempotencyKey: string): Promise<ComplianceAlert> {
    return this.client.request("POST", "/v1/compliance/alerts", {
      body: { ...data, tenantId: this.client.requireTenantId() },
      headers: { "Idempotency-Key": idempotencyKey },
    });
  }

  get(alertId: string): Promise<ComplianceAlert> {
    return this.client.request("GET", `/v1/compliance/alerts/${alertId}`, {
      query: { tenant_id: this.client.requireTenantId() },
    });
  }

  list(query: ListComplianceAlertsQuery = {}): Promise<ListComplianceAlertsResponse> {
    return this.client.request("GET", "/v1/compliance/alerts", {
      query: {
        tenant_id: this.client.requireTenantId(),
        subject_id: query.subjectId,
        status: query.status,
        severity: query.severity,
        limit: query.limit,
        cursor: query.cursor,
      },
    });
  }

  resolve(alertId: string, resolveNote: string): Promise<ComplianceAlert> {
    return this.client.request("POST", `/v1/compliance/alerts/${alertId}/resolve`, {
      body: { tenant_id: this.client.requireTenantId(), alert_id: alertId, resolve_note: resolveNote },
    });
  }

  escalate(alertId: string, escalatedTo: string, escalateNote: string): Promise<ComplianceAlert> {
    return this.client.request("POST", `/v1/compliance/alerts/${alertId}/escalate`, {
      body: {
        tenant_id: this.client.requireTenantId(),
        alert_id: alertId,
        escalated_to: escalatedTo,
        escalate_note: escalateNote,
      },
    });
  }

  async hasOpenAlerts(subjectId: string): Promise<boolean> {
    const response = await this.list({ subjectId, status: AlertStatus.Open, limit: 1 });
    return response.alerts.length > 0;
  }
}
